Shared library git ref

michaelsauter opened the pull request:
Addresses #773.

The main goal is to be able to use a Git tag for `ODS_GIT_REF` to allow admins to nail down exactly what version of ODS they install.

To achieve this, the PR also includes the following required changes:

* For users, we only want to communicate the "ODS version" as configured by `ODS_IMAGE_TAG` (which can be as precise or relaxed as admins want, e.g. it can be `4.x`, `4.0.x` or `4.0.0`). Therefore, this PR creates a `make` target that ensures a Git ref equal to `ODS_IMAGE_TAG` exists in `ods-jenkins-shared-library` so that this reference can be consumed by users from their `Jenkinsfile`. To finalise this particular change, we need another PR in the shared library to expose the value of `ODS_IMAGE_TAG` as `agentImageTag` and `sharedLibraryRef` to quickstarters (https://github.com/opendevstack/ods-jenkins-shared-library/pull/457), and finally a PR in the quickstarters to use `agentImageTag` and `sharedLibraryRef`
* Adjust the `ods-setup/repos.sh` script to handle Git tags properly. The PR achieves this by reworking `ods-setup/repos.sh`, which has been split into two scripts. The first script, `scripts/repos.sh` is supposed to be used from the umbrella directory to initialise and update local repos (from GitHub). The script is now safe to be used for initial installation and update using the exact same parameters. The second script, `scripts/push-local-repos.sh`, can be used to push those local repos to Bitbucket. This split makes each script easier to understand, both for ODS developers but also from a user perspective (e.g. because there are fewer parameters now)
* Adjust the ods-devenv box creation to make use of the new scripts and set the shared lib ref

clemensutschig commented:
@michaelsauter  - cool, but this has very likely major impact also on ods-devenv / deploy.sh or?

michaelsauter commented:
@clemensutschig Yes this affects ods-devenv, though only slightly. I adjusted the `deploy.sh` script (and also the corresponding installation docs), however I have not tested this part yet. @georgfedermann is this easy for you to run end-to-end as an AMI build? At this stage, it should be "just the same as before" and when we merge the PR in the shared lib / the quickstarters, then we are able to use a Git tag as the ODS source, and have a "virtual identifier" such as `4.x` as the image tag / shared lib ref for users to consume.

michaelsauter commented:
Marked this PR ready for review now. I adjusted the description above to clearly state the goal and the included changes.

@metmajer @henrjk @segator Pinging you here if you also have comments / concerns / feedback regarding this.

henrjk commented on scripts/repos.sh:
```diff
@@ -0,0 +1,160 @@
+    exit 1
+fi
+
+echo_done(){
```
Could add a comment that since this is used for bootstrapping it is not using colored_output.sh

michaelsauter replied:
Added as a comment in 10d7e94.

henrjk commented on scripts/repos.sh:
```diff
@@ -0,0 +1,160 @@
+#!/usr/bin/env bash
```
The script was moved, but in the PR I see it as delete and add which makes it harder to compare. Is it worthwhile to diff and review in detail?

michaelsauter replied:
It was moved, yes, but it has also been slimmed down considerably. Previously it was potentially also pushing (syncing) to Bitbucket, whereas now it is just clone/update from a remote. The general approach hasn't been changed though so I think you don't need to review in detail.

henrjk reviewed (commented):
This looks good to me and I like the much clearer installation adocs.
If needed I can take a more detailed look at `repos.sh`(see comment about this)

michaelsauter commented:
@henrjk Thanks for loving at this! I addressed your comments. Could you approve?

@segator @felipecruz91 Talking this over with you now, could you also approve from your side?

@georgfedermann Would you be able to check the part that touches the AMI build? That would be great.

clemensutschig commented:
@georgfedermann - merge it into ods-devenv

Michael Sauter <notifications@github.com> schrieb am Mo., 26. Okt. 2020,
17:07:

> @henrjk <https://github.com/henrjk> Thanks for loving at this! I
> addressed your comments. Could you approve?
>
> @segator <https://github.com/segator> @felipecruz91
> <https://github.com/felipecruz91> Talking this over with you now, could
> you also approve from your side?
>
> @georgfedermann <https://github.com/georgfedermann> Would you be able to
> check the part that touches the AMI build? That would be great.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/opendevstack/ods-core/pull/831#issuecomment-716651014>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJV7CSARJA4XESF5IJE5YA3SMWNE3ANCNFSM4SIK6ZEQ>
> .
>

clemensutschig commented:
@michaelsauter  - did you merge this into ods-devenv?

michaelsauter commented:
@clemensutschig No, I did not hear back from @georgfedermann yet. Anyway, I decided it is finally time to get some first-hand experience building this AMI so I'm setting up access to an AWS account at the moment and then I'll check for myself :)

georgfedermann commented:
ah, I thought I should also merge it into feature/ods-devenv as soon as it's approved. I did not know that it was about testing the change. Sorry about that. @michaelsauter can we create a PR against feature/ods-devenv?

georgfedermann commented:
sorry for the delay, this will be on the midnight build for feature/ods-devenv branch.

clemensutschig commented:
@michaelsauter  - looks like ods-devenv broke / can you work with @georgfedermann  to see whats going on?

georgfedermann commented:
@michaelsauter @clemensutschig the current build breaking issue was that packer would execute an outdated bootstrap.sh script. Sorry bout that, I fixed it for the 18:00 build.

georgfedermann commented:
@michaelsauter
After the fix of the packer scripts, the build for feature/ods-devenv is still failing.
After line repos.sh:125 the branch GIT_REF will not be checked-out / available locally for repository ods-document-generation-templates, and thus the build will break in push-local-repos.sh:100, when it tries to push feature/ods-devenv to bitbucket.

michaelsauter commented:
@georgfedermann Thanks for the detailed explanation. I missed to move the special handling of `ods-document-generation-templates` to `push-local-repos.sh`. Fixed in e329fb3.

georgfedermann commented:
@michaelsauter 
works! Please merge
